pgrx/src/nullable.rs

```diff
@@ -0,0 +1,466 @@
+    /// while `and_then()` can be used to reject a valid value and return
+    /// null anyway.
+    #[inline]
+    pub fn and_then<U, F>(self, func: F) -> Nullable<U>
+    where
+        F: FnOnce(T) -> Nullable<U>,
+    {
+        match self {
+            Nullable::Valid(v) => func(v),
+            Nullable::Null => Nullable::Null,
+        }
+    }
+}
+
+impl<T> Into<Option<T>> for Nullable<T> {
+    #[inline]
+    fn into(self) -> Option<T> {
+        match self {
+            Nullable::Valid(val) => Some(val),
+            Nullable::Null => None,
+        }
+    }
+}
+impl<T> Into<Nullable<T>> for Option<T> {
+    #[inline]
+    fn into(self) -> Nullable<T> {
+        match self {
+            Some(val) => Nullable::Valid(val),
+            None => Nullable::Null,
+        }
+    }
+}
+
+/* A null-bitmap array of (from postgres' perspective) length 8 with 3 nulls
+in it has an actual data array size of `size_of<Elem>() * 5` (give or
+take padding), whereas the bool array nulls implementation will still
+have an underlying data array of `size_of<Elem>() * 8` (give or take
+padding) no matter how many nulls there are. */
+// For the bitmap array, 1 is "valid", 0 is "null",
+// for the bool array,  1 is "null", 0 is "valid".
+
+/// Represents Postgres internal types which describe the layout of where
+/// filled slots (equivalent to Some(T)) are and where nulls (equivalent
+/// to None) are in the array.
+/// Note that a NullLayout must also capture the length of the container it
+/// refers to.
+pub trait NullLayout<Idx>
+where
+    Idx: PartialEq + PartialOrd,
+{
+    /// Returns true if this container has any nulls, checked at run-time.
+    /// # Performance
+    /// Different implementors of this type might have wildly varying
+    /// performance costs for checking this - for a null-bitslice-style
+    /// container, it should complete very quickly, but for a
+    /// null-boolean-array-style implementation, it could be O(n),
+    /// where n is the length of the container.
+    fn contains_nulls(&self) -> bool;
+
+    fn count_nulls(&self) -> usize;
+
+    /// Returns Some(true) if the element at `idx` is valid (non-null),
+    /// Some(false) if the element at `idx` is null,
+    /// or `None` if `idx` is out-of-bounds.
+    fn is_valid(&self, idx: Idx) -> Option<bool>;
+
+    /// Returns Some(true) if the element at idx is null,
+    /// Some(false) if the element at `idx` is valid (non-null),
+    /// or `None` if `idx` is out-of-bounds.
+    fn is_null(&self, idx: Idx) -> Option<bool> {
+        self.is_valid(idx).map(|v| !v)
+    }
+}
+
+pub trait SkippingNullLayout<Idx>: NullLayout<Idx>
+where
+    Idx: PartialEq + PartialOrd,
+{
+}
+
+/// All non-skipping null layouts. Marker trait for nullable containers that
+/// are nonetheless safe to iterate over linearly.
+pub trait ContiguousNullLayout<Idx>: NullLayout<Idx>
+where
+    Idx: PartialEq + PartialOrd,
+{
+}
```

Comment:
I don't think we need these marker traits, as they should be fully simply described by how the implementer implements the trait.